compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts

```diff
@@ -24,6 +35,341 @@ import {
   reversePostorderBlocks,
 } from '../HIR/HIRBuilder';
 import {CompilerError, EnvironmentConfig} from '..';
+import {
+  mapInstructionLValues,
+  mapInstructionOperands,
+  mapInstructionValueOperands,
+  mapTerminalOperands,
+} from '../HIR/visitors';
+
+type InlinedJsxDeclarationMap = Map<
+  DeclarationId,
+  {identifier: Identifier; blockIdsToIgnore: Set<BlockId>}
+>;
+
+/**
+ * A prod-only, RN optimization to replace JSX with inlined ReactElement object literals
+ *
+ * Example:
+ * <>foo</>
+ * _______________
+ * let t1;
+ * if (__DEV__) {
+ *   t1 = <>foo</>
+ * } else {
+ *   t1 = {...}
+ * }
+ *
+ */
+export function inlineJsxTransform(
+  fn: HIRFunction,
+  inlineJsxTransformConfig: NonNullable<
+    EnvironmentConfig['inlineJsxTransform']
+  >,
+): void {
+  const inlinedJsxDeclarations: InlinedJsxDeclarationMap = new Map();
+  /**
+   * Step 1: Codegen the conditional and ReactElement object literal
+   */
+  for (const [_, currentBlock] of [...fn.body.blocks]) {
+    let fallthroughBlockInstructions: Array<Instruction> | null = null;
+    const instructionCount = currentBlock.instructions.length;
+    for (let i = 0; i < instructionCount; i++) {
+      const instr = currentBlock.instructions[i]!;
+      switch (instr.value.kind) {
+        case 'JsxExpression':
+        case 'JsxFragment': {
+          /**
+           * Split into blocks for new IfTerminal:
+           *   current, then, else, fallthrough
+           */
+          const currentBlockInstructions = currentBlock.instructions.slice(
+            0,
+            i,
+          );
+          const thenBlockInstructions = currentBlock.instructions.slice(
+            i,
+            i + 1,
+          );
+          const elseBlockInstructions: Array<Instruction> = [];
+          fallthroughBlockInstructions ??= currentBlock.instructions.slice(
+            i + 1,
+          );
+
+          const fallthroughBlockId = fn.env.nextBlockId;
+          const fallthroughBlock = {
+            ...currentBlock,
+            id: fallthroughBlockId,
+            instructions: fallthroughBlockInstructions,
```

Comment:
probably safer to reset phis, preds, etc here, or more generally to just copy the specific properties you want to keep (which is really only block.kind)

Reply:
Updated to do explicit copies. I think we want to reuse the terminal from the current block in addition to the kind since we are splitting here?